crates/prune/src/prune.rs

```diff
@@ -0,0 +1,154 @@
+//! Support for maintaining the state of the transaction pool
+
+use futures_util::{Stream, StreamExt};
+use reth_primitives::BlockNumber;
+use reth_provider::CanonStateNotification;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{ready, Context, Poll},
+};
+use tracing::debug;
+
+// TODO(alexey): use config field from https://github.com/paradigmxyz/reth/pull/3341
+/// Minimum pruning interval measured in blocks. All prune parts are checked and, if needed, pruned,
+/// when the chain advances by the specified number of blocks.
+const MIN_PRUNE_BLOCK_INTERVAL: u64 = 5;
+
+/// Pruning routine. Implements [Future] where the pruning logic happens.
+pub struct Pruner<St, Client> {
+    /// Stream of canonical state notifications. Pruning is triggered by new incoming
+    /// notifications.
+    canon_state_stream: St,
+    /// Database interaction client.
+    #[allow(dead_code)]
+    client: Client,
+    /// Maximum reorg depth. Used to determine the pruning target for parts that are needed during
+    /// the reorg, e.g. changesets.
+    #[allow(dead_code)]
+    max_reorg_depth: u64,
+    /// Last pruned block number. Used in conjuction with [MIN_PRUNE_BLOCK_INTERVAL] to determine
+    /// when the pruning needs to be initiated.
+    last_pruned_block_number: Option<BlockNumber>,
+}
+
+impl<St, Client> Pruner<St, Client>
+where
+    St: Stream<Item = CanonStateNotification> + Send + Unpin + 'static,
+    Client: Send + Unpin + 'static,
+{
+    /// Creates a new [Pruner].
+    pub fn new(canon_state_stream: St, client: Client, max_reorg_depth: u64) -> Self {
+        Self { canon_state_stream, client, max_reorg_depth, last_pruned_block_number: None }
+    }
+}
+
+impl<St, Client> Future for Pruner<St, Client>
+where
+    St: Stream<Item = CanonStateNotification> + Send + Unpin + 'static,
+    Client: Send + Unpin + 'static,
+{
```

Comment:
the concept is actually quite simple:
listen for new blocks, check height, prune accordingly

Reply:
correct. added a tldr comment

Reply:
> listen for new blocks, check height, prune accordingly

given that static files will work in a very similar manner, could we abstract this in a manner that both pruning/static would share some functionality and/or interfaces?

Reply:
I think the `CanonStateNotification` is perfectly fine for this, we already use this in other places like:

https://github.com/paradigmxyz/reth/blob/6b8b478c6f6bf2d48e0c22d3437b7af47ef6290a/crates/transaction-pool/src/maintain.rs#L21-L36

and 

https://github.com/paradigmxyz/reth/blob/6b8b478c6f6bf2d48e0c22d3437b7af47ef6290a/crates/rpc/rpc/src/eth/cache.rs#L515-L520

Reply:
> could we abstract this in a manner that both pruning/static would share some functionality and/or interfaces? 

imo common functionality in pruning and snapshots is pretty thin: future that listens to new canon state notifications and then does implementation-specific logic. So not sure it worths abstracting it.